File: src/BridgePatternExample.cpp
#include <iostream>
#include <memory>

class Color {
public:
    virtual void applyColor() = 0;
};

class Red : public Color {
public:
    void applyColor() override {
        std::cout << "Red" << std::endl;
    }
};

class Green : public Color {
public:
    void applyColor() override {
        std::cout << "Green" << std::endl;
    }
};

class Shape {
public:
    virtual void draw() = 0;
};

class Circle : public Shape {
private:
    std::shared_ptr<Color> color;

public:
    Circle(std::shared_ptr<Color> color) : color(color) {}

    void draw() override {
        std::cout << "Drawing Circle in ";
        color->applyColor();
    }

    void setColor(std::shared_ptr<Color> newColor) {
        color = newColor;
    }
};

class Square : public Shape {
private:
    std::shared_ptr<Color> color;

public:
    Square(std::shared_ptr<Color> color) : color(color) {}

    void draw() override {
        std::cout << "Drawing Square in ";
        color->applyColor();
    }

    void setColor(std::shared_ptr<Color> newColor) {
        color = newColor;
    }
};

int main() {
    std::shared_ptr<Color> red = std::make_shared<Red>();
    std::shared_ptr<Color> green = std::make_shared<Green>();

    std::shared_ptr<Circle> circle = std::make_shared<Circle>(red);
    circle->draw();

    std::shared_ptr<Square> square = std::make_shared<Square>(green);
    square->draw();

    circle->setColor(green);
    std::cout << "Changing Circle color to ";
    circle->draw();

    return 0;
}

File: src/BridgePatternExample.cs
using System;

public interface Color
{
    void applyColor();
}

public class Red : Color
{
    public void applyColor()
    {
        Console.WriteLine("Red");
    }
}

public class Green : Color
{
    public void applyColor()
    {
        Console.WriteLine("Green");
    }
}

public interface Shape
{
    void draw();
}

public class Circle : Shape
{
    private Color color;

    public Circle(Color color)
    {
        this.color = color;
    }

    public void draw()
    {
        Console.Write("Drawing Circle in ");
        color.applyColor();
    }

    public void setColor(Color newColor)
    {
        color = newColor;
    }
}

public class Square : Shape
{
    private Color color;

    public Square(Color color)
    {
        this.color = color;
    }

    public void draw()
    {
        Console.Write("Drawing Square in ");
        color.applyColor();
    }

    public void setColor(Color newColor)
    {
        color = newColor;
    }
}

class BridgePatternExample
{
    static void Main(string[] args)
    {
        Color red = new Red();
        Color green = new Green();

        Circle circle = new Circle(red);
        circle.draw();

        Square square = new Square(green);
        square.draw();

        circle.setColor(green);
        Console.Write("Changing Circle color to ");
        circle.draw();
    }
}
